File: Minsk/SyntaxNode.cs
namespace Minsk;

public abstract class SyntaxNode
{
    public abstract SyntaxKind Kind { get; }
    public abstract IEnumerable<SyntaxNode> Children { get; }

    public void PrettyPrint()
    {
        PrettyPrintInternal(this, "", true);
    }

    private static void PrettyPrintInternal(SyntaxNode node, string indent, bool isLast)
    {
        Console.Write(indent);
        var marker = isLast ? "└───" : "├───";
        Console.Write(marker);
        if (node is SyntaxToken t)
        {
            Console.Write(t);
        }
        else
        {
            Console.Write(node.Kind);
        }

        Console.WriteLine();

        indent += isLast ? "    " : "│   ";
        var children = node.Children.ToArray();
        var lastChild = children.LastOrDefault();
        foreach (var child in children)
        {
            PrettyPrintInternal(child, indent, child == lastChild);
        }
    }
}

File: kt/Minsk/src/main/kotlin/dev/phytolizer/minsk/SyntaxNode.kt
package dev.phytolizer.minsk

abstract class SyntaxNode {
    abstract val kind: SyntaxKind
    abstract val children: List<SyntaxNode>

    fun prettyPrint() {
        prettyPrint(this, "", true)
    }

    private fun prettyPrint(node: SyntaxNode, indent: String, isLast: Boolean) {
        print(indent)
        val marker = if (isLast) {
            "└── "
        } else {
            "├── "
        }
        print(marker)
        if (node is SyntaxToken) {
            print(node)
        } else {
            print(node.kind)
        }
        println()
        val newIndent = indent + if (isLast) {
            "    "
        } else {
            "│   "
        }
        val lastChild = node.children.lastOrNull()
        for (child in node.children) {
            prettyPrint(child, newIndent, child == lastChild)
        }
    }
}
